File: c++-srcs/logic/dd/gtest/BddMgr_test.cc

/// @file BddMgr_test.cc
/// @brief BddMgr のユニットテスト
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2022, 2023 Yusuke Matsunaga
/// All rights reserved.

#include <gtest/gtest.h>
#include "ym/Bdd.h"
#include "ym/BddMgr.h"
#include "ym/BddVar.h"
#include "ym/Expr.h"
#include "BddTest.h"


BEGIN_NAMESPACE_YM

TEST_F(BddTest, zero)
{
  auto bdd = Bdd::zero();

  EXPECT_TRUE( bdd.is_zero() );
  EXPECT_FALSE( bdd.is_one() );
  EXPECT_TRUE( bdd.is_const() );
  EXPECT_FALSE( bdd.is_invalid() );

  check(bdd, "0");
}

TEST_F(BddTest, one)
{
  auto bdd = Bdd::one();

  EXPECT_FALSE( bdd.is_zero() );
  EXPECT_TRUE( bdd.is_one() );
  EXPECT_TRUE( bdd.is_const() );
  EXPECT_FALSE( bdd.is_invalid() );

  check(bdd, "1");
}

TEST_F(BddTest, variable1)
{
  auto var = mMgr.variable(0);
  EXPECT_FALSE( var.is_invalid() );

  auto bdd = var.bdd();
  EXPECT_FALSE( bdd.is_zero() );
  EXPECT_FALSE( bdd.is_one() );
  EXPECT_FALSE( bdd.is_const() );
  EXPECT_FALSE( bdd.is_invalid() );

  check(bdd, "10");

  Bdd f0;
  Bdd f1;
  auto var1 = bdd.root_decomp(f0, f1);

  EXPECT_EQ( var1, var );
  EXPECT_TRUE( f0.is_zero() );
  EXPECT_TRUE( f1.is_one() );
}

TEST_F(BddTest, from_truth1)
{
  const char* exp_str = "10010110";
  Bdd bdd = mMgr.from_truth(exp_str);
  check(bdd, exp_str);
}

TEST_F(BddTest, from_truth2)
{
  const char* exp_str = "10010101";
  Bdd bdd = mMgr.from_truth(exp_str);
  check(bdd, exp_str);
}

TEST_F(BddTest, from_truth3)
{
  auto var0 = mMgr.variable(0);
  auto var1 = mMgr.variable(1);
  auto var2 = mMgr.variable(2);
  std::vector<BddVar> var_list{var2, var1, var0};
  const char* exp_str = "10010101";
  Bdd bdd = mMgr.from_truth(exp_str, var_list);
  check(bdd, exp_str);
}

TEST_F(BddTest, and1)
{
  auto var0 = mMgr.variable_bdd(0);
  auto var1 = mMgr.variable_bdd(1);
  auto f = var0 & var1;
  check(f, "1000");
}

TEST_F(BddTest, from_expr1)
{
  auto expr = Expr::zero();
  auto bdd = mMgr.from_expr(expr);
  check(bdd, "0");
}

TEST_F(BddTest, from_expr2)
{
  auto expr = Expr::one();
  auto bdd = mMgr.from_expr(expr);
  check(bdd, "1");
}

TEST_F(BddTest, from_expr3)
{
  auto expr = Expr::positive_literal(0);
  auto bdd = mMgr.from_expr(expr);
  check(bdd, "10");
}

TEST_F(BddTest, from_expr4)
{
  auto expr = Expr::negative_literal(0);
  auto bdd = mMgr.from_expr(expr);
  check(bdd, "01");
}

TEST_F(BddTest, from_expr5)
{
  auto expr1 = Expr::positive_literal(0);
  auto expr2 = Expr::positive_literal(1);
  auto bdd = mMgr.from_expr(expr1 & expr2);
  check(bdd, "1000");
}

TEST_F(BddTest, from_expr6)
{
  auto expr1 = Expr::positive_literal(0);
  auto expr2 = Expr::positive_literal(1);
  auto bdd = mMgr.from_expr(expr1 | expr2);
  check(bdd, "1110");
}

TEST_F(BddTest, from_expr7)
{
  auto expr1 = Expr::positive_literal(0);
  auto expr2 = Expr::positive_literal(1);
  auto bdd = mMgr.from_expr(expr1 ^ expr2);
  check(bdd, "0110");
}

TEST_F(BddTest, copy)
{
  const char* exp_str = "1001010110111010";
  Bdd bdd = mMgr.from_truth(exp_str);

  BddMgr mgr1;
  Bdd bdd1 = mgr1.copy(bdd);
  EXPECT_TRUE( !bdd1.is_invalid() );

  auto bdd2 = mMgr.copy(bdd1);
  check(bdd2, exp_str);
}

TEST_F(BddTest, mgr_copy)
{
  BddMgr mgr1;
  auto var = mgr1.variable(0);

  auto bdd = var.positive_literal();

  auto mgr2 = bdd.mgr();

  EXPECT_EQ( mgr1, mgr2 );
}

TEST_F(BddTest, mgr_lifetime)
{
  Bdd bdd;
  {
    BddMgr mgr;
    auto var = mgr.variable(0);
    auto lit = var.positive_literal();
    bdd = lit.bdd();
  }

  //check(bdd, "10");
}

TEST_F(BddTest, mgr_var_ordering)
{
  auto var0 = variable(0);
  auto var1 = variable(1);
  auto var2 = variable(2);
  auto var3 = variable(3);

  EXPECT_EQ( 0, var0.varid() );
  EXPECT_EQ( 1, var1.varid() );
  EXPECT_EQ( 2, var2.varid() );
  EXPECT_EQ( 3, var3.varid() );

  auto v_list = mMgr.variable_list();
  std::vector<BddVar> exp_v_list{var0, var1, var2, var3};
  EXPECT_EQ( exp_v_list, v_list );

  mMgr.set_variable_order({var3, var2, var1, var0});

  auto o_list = mMgr.variable_order();
  std::vector<BddVar> exp_o_list{var3, var2, var1, var0};
  EXPECT_EQ( exp_o_list, o_list );
}

TEST_F(BddTest, mgr_bdd_size1)
{
  auto bdd1 = from_truth("1000");
  auto bdd2 = from_truth("1110");

  EXPECT_EQ( 3, bdd_size({bdd1, bdd2}) );
}

TEST_F(BddTest, mgr_display1)
{
  auto var0 = variable_bdd(0);
  auto var1 = variable_bdd(1);

  auto bdd1 = var0 & var1;
  auto bdd2 = var0 | var1;

  std::ostringstream os;
  display(os, {bdd1, bdd2});

  static const char* exp_str =
    "     2      3 \n"
    "     3:    0     1 :     ONE\n"
    "     2:    0   ZERO:      1 \n"
    "     1:    1   ZERO:     ONE\n";

  EXPECT_EQ( exp_str, os.str() );
}

TEST_F(BddTest, mgr_display2)
{
  auto var0 = variable(0);
  auto var1 = variable(1);

  mMgr.set_variable_order({var1, var0});

  auto f0 = var0.bdd();
  auto f1 = var1.bdd();
  auto bdd1 = f0 & f1;
  auto bdd2 = f0 | f1;

  std::ostringstream os;
  display(os, {bdd1, bdd2});

  static const char* exp_str =
    "     2      3 \n"
    "     3:    1     1 :     ONE\n"
    "     2:    1   ZERO:      1 \n"
    "     1:    0   ZERO:     ONE\n";

  EXPECT_EQ( exp_str, os.str() );
}

END_NAMESPACE_YM

File: py_capi/pytest/logic/bddmgr_test.py
#! /usr/bin/env python3

""" BddMgr のテストプログラム

:file: bddmgr_test.py
:author: Yusuke Matsunaga (松永 裕介)
:copyright: Copyright (C) 2024 Yusuke Matsunaga, All rights reserved.
"""

import pytest
from ymworks.logic import Bdd, BddMgr
from ymworks.logic import Expr
from check_bdd import check


def test_variable1():
    mgr = BddMgr()

    var = mgr.variable(0)
    assert not var.is_invalid()
    assert var.varid == 0

    bdd = var.bdd()
    assert not bdd.is_invalid()
    assert not bdd.is_zero()
    assert not bdd.is_one()

    bdd2 = mgr.variable_bdd(0)
    assert bdd2 == bdd

    var1, f0, f1 = bdd.root_decomp()

    assert var1 == var
    assert f0.is_zero()
    assert f1.is_one()

    assert check(bdd, mgr, "10")

def test_variable2():
    mgr = BddMgr()

    var = mgr.variable(1)
    assert not var.is_invalid()
    assert var.varid == 1

    bdd = ~var.bdd()
    assert not bdd.is_invalid()
    assert not bdd.is_zero()
    assert not bdd.is_one()

    bdd2 = ~mgr.variable_bdd(1)
    assert bdd2 == bdd

    var1, f0, f1 = bdd.root_decomp()

    assert var1 == var
    assert f0.is_one()
    assert f1.is_zero()

    assert check(bdd, mgr, "0101")

def test_variable3():
    mgr = BddMgr()

    var = mgr.variable(2)
    assert not var.is_invalid()
    assert var.varid == 2

    bdd = var.bdd()
    assert not bdd.is_invalid()
    assert not bdd.is_zero()
    assert not bdd.is_one()

    bdd2 = mgr.variable_bdd(2)
    assert bdd2 == bdd

    var1, f0, f1 = bdd.root_decomp()

    assert var1 == var
    assert f0.is_zero()
    assert f1.is_one()

    assert check(bdd, mgr, "10101010")

def test_variable4():
    mgr = BddMgr()

    var = mgr.variable(3)
    assert not var.is_invalid()
    assert var.varid == 3

    bdd = ~var.bdd()
    assert not bdd.is_invalid()
    assert not bdd.is_zero()
    assert not bdd.is_one()

    bdd2 = ~mgr.variable_bdd(3)
    assert bdd2 == bdd

    var1, f0, f1 = bdd.root_decomp()

    assert var1 == var
    assert f0.is_one()
    assert f1.is_zero()

    assert check(bdd, mgr, "0101010101010101")

def test_from_truth1():
    mgr = BddMgr()

    exp_str = "10010110"
    bdd = mgr.from_truth(exp_str)

    assert not bdd.is_invalid()

    assert check(bdd, mgr, exp_str)

def test_from_truth2():
    mgr = BddMgr()

    exp_str = "10010101"
    bdd = mgr.from_truth(exp_str)

    assert not bdd.is_invalid()

    assert check(bdd, mgr, exp_str)

def test_from_truth3():
    mgr = BddMgr()

    var0 = mgr.variable(0)
    var1 = mgr.variable(1)
    var2 = mgr.variable(2)
    var_list = ( var2, var1, var0 )
    exp_str = "10010101"
    bdd = mgr.from_truth(exp_str,
                         var_list=var_list)

    assert not bdd.is_invalid()

    assert check(bdd, mgr, exp_str)

def test_from_expr1():
    mgr = BddMgr()

    var_list = [ mgr.variable(i) for i in range(3) ]
    expr_str = "0 + (1 * ~2)"
    expr = Expr.from_string(expr_str)
    bdd = mgr.from_expr(expr,
                        var_list=var_list)

    assert not bdd.is_invalid()

    print(bdd.to_truth(var_list))

def test_copy():
    mgr1 = BddMgr()

    var_list = [ mgr1.variable(i) for i in range(3) ]
    exp_str = "10010110"
    bdd = mgr1.from_truth(exp_str,
                          var_list=var_list)

    mgr2 = BddMgr()
    bdd2 = mgr2.copy(bdd)

    assert check(bdd2, mgr2, exp_str)
